### PR TITLE
EditorService, TaskController, TaskImplementationService

### PR DESCRIPTION
**Pull request by lramirez**
* EditorService generates tab contents
* TaskController refactored
* DataImporter improved
#  do not merge yet, cleaning up

### CODE REVIEW

**Comment by dwarfcrank** [comment]
I think a bigger issue than the size of the PR is that neither the title nor the description actually say what the PR does...

File: build.gradle [hunk 1]
```diff
@@ -1,7 +1,10 @@
 buildscript {
     repositories {
         mavenCentral()
-        maven { url "https://repo.spring.io/milestone" }
+        maven {
+            url "https://repo.spring.io/milestone"
+            //url "https://plugins.gradle.org/m2/"
```
Comment (laurpulk) [inline build.gradle]: Try to remove unnecessary comments before doing PR

File: build.gradle [hunk 2]
```diff
@@ -109,6 +113,34 @@ bootRun {
...
+//    doFirst {
+//        generatedSourcesDir.mkdirs()
+//    }
+//    dependsOn(removeGeneratedMetamodel)
+//    options.compilerArgs += ['-s', generatedSourcesDir]
+//}
+
+//clean.dependsOn(removeGeneratedMetamodel)
+
+
```
Comment (laurpulk) [inline build.gradle]: // :-)

File: config/application.yml [hunk 3]
```diff
@@ -11,7 +11,7 @@ spring:
   profiles:
     default: dev
     active: oauth
-
+
```
Comment (laurpulk) [inline config/application.yml]: Try to avoid adding blank lines if not necessary, try using "**git add -p**" instead of "_git add ._"

File: src/main/java/pingis/controllers/UserDevController.java [hunk 4]
```diff
@@ -17,6 +17,7 @@ public String login(Model model) {
 
     @RequestMapping(value = "/user", method = RequestMethod.GET)
     public String user(Model model) {
+
```
Comment (laurpulk) [inline src/main/java/pingis/controllers/UserDevController.java:20]: git add -p

File: src/main/java/pingis/entities/Challenge.java [hunk 5]
```diff
@@ -51,6 +51,7 @@
     
     protected Challenge() {}
 
+
```
Comment (laurpulk) [inline src/main/java/pingis/entities/Challenge.java:54]: git add -p

File: src/main/java/pingis/entities/OldSpecifications.java [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+                return cb.equal(root.get(Task_.challenge), challenge);
+            }
+        };
+    }
+    
+   
+        
+  
+}
+*/
```
Comment (laurpulk) [inline src/main/java/pingis/entities/OldSpecifications.java]: Unnecessary file

File: src/main/java/pingis/entities/TaskImplementation.java [hunk 7]
```diff
@@ -109,4 +109,5 @@ public void setChallengeImplementation(ChallengeImplementation challengeImplemen
     public ChallengeImplementation getChallengeImplementation() {
         return challengeImplementation;
     }
+
```
Comment (laurpulk) [inline src/main/java/pingis/entities/TaskImplementation.java:112]: git add -p

File: src/main/java/pingis/repositories/ChallengeImplementationRepository.java [hunk 8]
```diff
@@ -2,6 +2,7 @@
  */
 package pingis.repositories;
 
+
```
Comment (laurpulk) [inline src/main/java/pingis/repositories/ChallengeImplementationRepository.java:5]: git add -p

File: src/main/java/pingis/repositories/ChallengeRepository.java [hunk 9]
```diff
@@ -1,9 +1,10 @@
 package pingis.repositories;
 
-import java.util.List;
 import pingis.entities.Challenge;
 import org.springframework.data.repository.CrudRepository;
 
 public interface ChallengeRepository extends CrudRepository<Challenge, Long> {
+
     Challenge findByName(String name);
+
```
Comment (laurpulk) [inline src/main/java/pingis/repositories/ChallengeRepository.java:9]: git add -p

File: src/main/java/pingis/services/DataLoader.java [hunk 10]
```diff
@@ -85,7 +85,7 @@ private void generateChallenges() {
         // Challenge 1        
         this.calculatorChallenge = new Challenge(
                 "Calculator", users.get((int) UserType.TMC_MODEL_USER.getId()), "Test and implement a simple text-based calculator.");
-        
+
```
Comment (laurpulk) [inline src/main/java/pingis/services/DataLoader.java:88]: git add -p

File: build.gradle [hunk 11]
```diff
@@ -120,7 +154,7 @@ cucumberOAuthTest {
 
 //Full debug prints for Travis, custom report prints for local
 if(System.env.TRAVIS == 'true') {
-  unitTest.testLogging.showStandardStreams = true
+  unitTest.testLogging.showStandardStreams = false
```
Comment (dwarfcrank) [inline build.gradle:123]: A fix for this is on the way, so this might not be needed in this PR.

File: config/application.yml [hunk 12]
```diff
@@ -10,8 +10,8 @@ spring:
     driverClassName: org.h2.Driver
   profiles:
     default: dev
-    active: oauth
-
+    active: dev
+
```
Comment (dwarfcrank) [inline config/application.yml]: Try not to commit local temporary changes to `application.yml`.

File: config/checkstyle/checkstyle.xml [hunk 13]
```diff
@@ -1,7 +1,6 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <!DOCTYPE module PUBLIC "-//Puppy Crawl//DTD Check Configuration 1.3//EN" "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
 <module name="Checker">
-
```
Comment (dwarfcrank) [inline config/checkstyle/checkstyle.xml:4]: These changes don't seem to be needed either.

File: src/main/java/pingis/controllers/UserController.java [hunk 14]
```diff
@@ -25,4 +25,4 @@ public String user(Model model) {
     public String admin(Model model) {
         return "admin";
     }
-}
+}
```
Comment (dwarfcrank) [inline src/main/java/pingis/controllers/UserController.java:28]: Unnecessary change.

File: src/main/java/pingis/services/DataImporter.java [hunk 15]
```diff
@@ -199,6 +214,30 @@ private Task createTask(ImplementationType type, User author, JSONObject taskObj
         task.setChallenge(challenge);
         return task;
     }
+    
+    private void createDummyImplementations(JSONArray implementations) {
+        for (int i = 0; i < implementations.length(); i++) {
+            JSONObject implementation = implementations.getJSONObject(i);
+            ChallengeImplementation challengeImplementation =
```
Comment (dwarfcrank) [inline src/main/java/pingis/services/DataImporter.java:221]: Try to add some whitespace between statements, makes it easier to follow.

File: src/main/java/pingis/services/IO.java [hunk 16]
```diff
@@ -5,10 +5,6 @@
  */
 package pingis.services;
 
-/**
```
Comment (dwarfcrank) [inline src/main/java/pingis/services/IO.java:8]: Unnecessary change. Changes like removing author comment blocks should be done in a separate PR.

File: src/test/java/pingis/controllers/TaskControllerTest.java [hunk 17]
```diff
@@ -53,84 +58,135 @@
...
                 .webAppContextSetup(context)
                 .build();
     }
 
+
+
     @Test
-    public void submitTask() throws Exception {
+    public void givenTaskWhenGetTestTask() throws Exception {
+        when(taskImplementationServiceMock.findOne(testTaskImplementation.getId()))
```
Comment (dwarfcrank) [inline src/test/java/pingis/controllers/TaskControllerTest.java:113]: Same here re: blank lines between statements or groups of statements.

File: src/test/java/pingis/controllers/TaskControllerTest.java [hunk 18]
```diff
@@ -53,84 +58,135 @@
...
+                .thenReturn(testTaskImplementation);
+        Map<String, EditorTabData> tabData = this.generateTestTabData(implTaskImplementation, testTaskImplementation);
+        when(editorServiceMock.generateEditorContents(testTaskImplementation)).thenReturn(tabData);
+        String uri = "/task/"+testTaskImplementation.getId();
+        performSimpleGetRequestAndFindContent(uri, "task", testTask.getCodeStub());
+        verify(taskImplementationServiceMock, times(1)).findOne(testTaskImplementation.getId());
+        verify(editorServiceMock, times(1))
```
Comment (dwarfcrank) [inline src/test/java/pingis/controllers/TaskControllerTest.java:120]: Calling `verify()` without a second parameter is equal to `verify(..., times(1))`, which can make lines shorter.

File: src/test/java/pingis/services/EditorServiceTest.java [hunk 19]
```diff
@@ -1,30 +1,92 @@
...
+import pingis.Application;
+import pingis.entities.Challenge;
+import pingis.entities.ChallengeImplementation;
 import pingis.entities.ImplementationType;
+import pingis.entities.TaskImplementation;
+import pingis.utils.EditorTabData;
+
 
+@RunWith(SpringRunner.class)
+@ContextConfiguration(classes = {Application.class})
```
Comment (dwarfcrank) [inline src/test/java/pingis/services/EditorServiceTest.java:26]: Does the whole application context need to be loaded for this test?

Comment (lramirez) [reply src/test/java/pingis/services/EditorServiceTest.java:26]: Apparently yes.
